### PR TITLE
allow override of constructGetUri along w/ potential CORS support

### PR DESCRIPTION
**Pull request by longlho**
cc @mridgway @lingyan

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by lingyan** [comment]
cc: @Vijar

File: libs/fetcher.client.js [hunk 1]
```diff
@@ -286,10 +288,14 @@ Queue.prototype = {
         // Helper Methods
         // ------------------------------------------------------------------
         /**
-         * @method _constructGetUri
-         * @private
+         * Construct GET URI. You can override this for custom GET URI construction
+         * @method constructGetUri
```
Comment (redonkulus) [inline libs/fetcher.client.js]: Should be `@defaultConstructGetUri`

File: libs/fetcher.client.js [hunk 2]
```diff
@@ -286,10 +288,14 @@ Queue.prototype = {
...
+         * @param {String} uri base URI
+         * @param {String} resource Resource name
+         * @param {params} Parameters to be serialized
```
Comment (redonkulus) [inline libs/fetcher.client.js]: Should be `@param {Object} params Parameters to be serialized`

File: libs/fetcher.client.js [hunk 3]
```diff
@@ -160,6 +160,8 @@ Queue.prototype = {
 
     function Fetcher (options) {
         this.xhrPath = options.xhrPath || DEFAULT_XHR_PATH;
+        this.corsPath = options.corsPath;
```
Comment (redonkulus) [inline libs/fetcher.client.js]: Both new params should be added to the doc block above.

**Comment by redonkulus** [comment]
@longlho do we have any tests we can update for this support?

**Comment by longlho** [comment]
@redonkulus  I'll definitely add tests, just wanna make sure that this use case is something we're willing to support :)

**Comment by longlho** [comment]
@redonkulus fixes & tests added :)

**Comment by redonkulus** [comment]
@longlho There are still other comments that still seem to be issues.

**Comment by longlho** [comment]
@redonkulus oh I decided to keep `constructGetUri` instead of `defaultConstructGetUri` and allow overrides per call instead since on `fluxible-plugin-fetchr` there seems to be only 1 Fetcher instance, thus having that override in constructor is not that useful. I added docs for `corsPath` too.

File: libs/util/http.client.js [hunk 4]
```diff
@@ -161,7 +163,9 @@ function io(url, options) {
         method: options.method || METHOD_GET,
         timeout: options.timeout,
         headers: options.headers,
-        body: options.data
+        body: options.data,
+        cors: options.cors,
+        useXDR: options.useXDR
```
Comment (Vijar) [inline libs/util/http.client.js]: Looks like `cors` and `useXDR` serve the same purpose for the `xhr` module. `useXDR` is documented while `cors` isn't, so I would use `useXDR` only.

File: README.md [hunk 5]
```diff
@@ -62,8 +62,34 @@ var Fetcher = require('fetchr'),
...
+```js
+var qs = require('qs');
+function constructGetUri (uri, resource, params, config) {
+	// this refers to the Fetcher object itself that this function
+	// is invoked with.
+	if (config.cors) {
+		return uri + '/' + resource + '?' + qs.stringify(this.context);
+	}
+	
+	return this.constructGetUri.apply(this, arguments);
```
Comment (Vijar) [inline README.md]: I don't really like letting the user knowing about and calling an internal function. Could we just have the user return a `falsy` value and we will automatically call the internal default function for them as fallback?

File: libs/util/http.client.js (hunk 4 above)
Comment (longlho) [inline libs/util/http.client.js]: :+1:

File: README.md (hunk 5 above)
Comment (longlho) [inline README.md]: :+1:

**Comment by longlho** [comment]
@Vijar fixes made :)

File: libs/fetcher.client.js [hunk 6]
```diff
@@ -286,8 +288,13 @@ Queue.prototype = {
...
         /**
+         * Construct GET URI. You can override this for custom GET URI construction
          * @method _constructGetUri
          * @private
+         * @param {String} uri base URI
+         * @param {String} resource Resource name
+         * @param {Object} params Parameters to be serialized
+         * @param {Object} Configuration object
          */
         _constructGetUri: function (uri, resource, params, config) {
```
Comment (Vijar) [inline libs/fetcher.client.js]: I'd want this to be `_defaultConstructGetUri` since that is what this fn is.

**Comment by Vijar** [comment]
@longlho once you make that final change, this looks good to me.

**Comment by longlho** [comment]
@Vijar changes made :)

**Comment by Vijar** [comment]
:ship:

**Comment by Vijar** [comment]
@redonkulus can you take one more look? and we can merge.

**Comment by redonkulus** [comment]
:+1:

**Comment by Vijar** [comment]
Thanks @longlho

**Comment by longlho** [comment]
awesome thanks guys!

**Pull request merged**